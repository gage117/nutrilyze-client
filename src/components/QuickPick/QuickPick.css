.quick-pick {
  width: 250px;
  min-height: 415px;
  border: 1px solid rgb(93, 84, 84);
  margin: 0px 0px 10px 10px;
}

.meals-header {
  background-color: #AAA;
  border-bottom: 1px solid black;
  padding: 5px;
  display: flex;
  justify-content: space-between;
}

.qp-h3 {
  margin: 0;
}

.ingredients-header {
  border-top: 1px solid black;
  background-color: #AAA;
  border-bottom: 1px solid black;
  padding: 5px;
}

.list-div {
  border-bottom: 1px solid #AAA;
}

.list-div > span {
  display: inline-block;
  padding: 2.5px 0px 2.5px 5px;
  margin: 0;
}

.hidden {
  display: none;
}

.serving-input {
  width: 40px;
}

.expand-div {
  background-color: #EEE;
  padding: 5px;
}

.ingredients-list {
  background-color: #EEE;
}

.ingredients-list > p {
  margin: 0px;
  padding: 0px 15px;
  font-weight: bold;
}

.ingredients-list > span {
  padding: 2.5px 0px 2.5px 25px;
  font-size: 16px;
}