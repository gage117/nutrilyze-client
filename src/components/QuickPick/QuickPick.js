import React from 'react'
import ReactDOM from 'react-dom'
import './QuickPick.css'

function QuickPick(props) {
  /** JSX Generators */
  function generateIngredient(ingredient, index) {
    return (
      <div key={index} className='list-div'>
        <div id={ingredient.name} className='qp__ingredient-bar qp__border-bottom'>
          <span>{ingredient.name}</span>
          <button id={`${ingredient.name}-button`} className={`${ingredient.name} qp__add-button`} onClick={handleAddClick}>Add</button>
        </div>
        <div id={`${ingredient.name}-hidden`} className='qp__hidden expand-div qp__border-bottom'></div>
      </div>
    )
  }
  /** Meal functionality */
  // function generateMeal(meal, index) {
  //   return (
  //     <div key={index} className='list-div'>
  //       <span>{meal.name}</span>
  //       <button className={meal.name} onClick={handleAddClick}>Add</button>
  //       <div id={meal.name} className='hidden expand-div'></div>
  //     </div>
  //   )
  // }
  function generateForm(item) {
    /** Multiple users functionality */
    /** Returns the checkbox and label JSX of a user */
    // function generateUserChoices(user) {
    //   return (
    //     <div key={user.id}>
    //       <input type='checkbox' className='user-checkbox' id={`${user.name}-list`} name={user.name} />
    //       <label htmlFor={`${user.name}-list`}>{user.name}</label>
    //     </div>
    //   )
    // }
    function getUnitOfMeasureByName(name) {
      const item = getItemByName(name)
      return item.unit_of_measure
    }
    /** Multiple users, Apply-to Checkboxes JSX, add above serving size div */
    // <div className='apply-to-half'>
    //   <span className='apply-to-span'>Apply To: </span>
    //   {props.state.users.map(user => generateUserChoices(user))}
    // </div>
    return (
      <form onSubmit={handleApplyClick} id={`${item}-apply`} className={`qp__add-form`}>
        <label className='qp__add-food-label'>
          <span className='block'>Serving:</span>
          <input type='text' id={`${item}-serving`} className='qp__serving-input' required />
          <span>{getUnitOfMeasureByName(item)}</span>
        </label>
        <button className='qp__apply-button'>Apply</button>
      </form>
    )
  }

  /** gets entire item object by name */
  function getItemByName(name) {
    let item;
    props.state.ingredients.map(ingredient => {
      if (ingredient.name === name) {
        item = ingredient
      }
      return item
    })
    /** Meals functionality */
    // props.state.meals.map(meal => {
    //   if (meal.name === name) {
    //     item = meal
    //   }
    //   return item
    // })
    return item
  }

  /**  */

  /** Button handlers */
  function handleApplyClick(event) {
    event.preventDefault()
    const item = getItemByName(event.target.getAttribute('id').split('-')[0])
    
    let serving_size = item.serving_size;
    if (isNaN(serving_size)) {
      serving_size = serving_size.split('/')
      serving_size = parseFloat(parseInt(serving_size[0]) / parseInt(serving_size[1])).toFixed(2)
    }
    console.log(serving_size)
    const multiplier = parseFloat(event.target[`${item.name}-serving`].value / Number(serving_size)).toFixed(2)
    
    if (isNaN(multiplier)) {
      alert('Serving size must be a number or a fraction (1/3)')
    } else {
      const nutritionValues = {
        calories: parseFloat(item.calories * multiplier).toFixed(0),
        protein: parseFloat(item.protein * multiplier).toFixed(2),
        carbs: parseFloat(item.carbs * multiplier).toFixed(2),
        sugar: parseFloat(item.sugar * multiplier).toFixed(2),
        fiber: parseFloat(item.fiber * multiplier).toFixed(2),
        fat: parseFloat(item.fat * multiplier).toFixed(2),
        sodium: parseFloat(item.sodium * multiplier).toFixed(2),
      }
      
      const usersToUpdate = [props.state.users[0]];
      /** Multiple users functionality */
      // for (let i = 0; i < props.state.users.length; i++) {
      //   if (event.target[`${props.state.users[i].name}-list`].checked && !usersToUpdate.includes(props.state.users[i])) {
      //     usersToUpdate.push(props.state.users[i])
      //   }
      // }
  
      props.handlers.updateUserNutrition(nutritionValues, usersToUpdate)
    }
  }
  function handleAddClick(event) {
    const item = event.target.getAttribute('id').split('-')[0]
    const element = generateForm(item)
    document.getElementById(`${item}-hidden`).classList.toggle('qp__hidden')
    document.getElementById(item).classList.toggle('qp__border-bottom')
    ReactDOM.render(element, document.getElementById(`${item}-hidden`))
  }
  /** Meal functionality */
  // function handleOrganizeMealClick() {

  // }
  /** Meal JSX, add to render above ingredients section */
  // <section className='qp-section meals'>
  //   <div className='meals-header'>
  //     <h3 className='qp-h3'>Meals</h3>
  //     <button>Organize New Meal +</button>
  //   </div>
  //   <div className='list-div'>
  //     <span>Spaghetti {'(HARD CODED)'}</span>
  //     <div className='meal-ingredients-list'>
  //       <p>> Ingredients {'(HARD CODED)'}</p>
  //       <span>Spaghetti Noodles, Pasta Sauce</span>
  //     </div>
  //   </div>
  //   {props.state.meals.map((meal, index) => generateMeal(meal, index))}
  // </section>

  return (
    <div className='quick-pick'>
      <section className='qp__section qp__ingredients'>
        <h2 className='qp__ingredients-header qp__h2'>Ingredients</h2>
        <div id='qp__ingredients-container'>
          {props.state.ingredients.map((ingredient, index) => generateIngredient(ingredient, index))}
        </div>
      </section>
    </div>
  )
}

export default QuickPick
