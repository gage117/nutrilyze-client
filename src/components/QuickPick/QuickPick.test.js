import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import {BrowserRouter} from 'react-router-dom'
import QuickPick from './QuickPick'

describe('QuickPick', () => {
  const state = {
    ingredients: [
      {
        name: 'Banana',
        calories: 89,
        protein: 1.1,
        carbs: 22.8,
        sugar: 12.2,
        fiber: 2.6,
        fat: 0.3,
        sodium: 0,
        serving_size: '100',
        unitOfMeasure: 'g'
      },
      {
        name: 'Broccoli',
        calories: 31,
        protein: 2.5,
        carbs: 6,
        sugar: 1.5,
        fiber: 2.4,
        fat: 0.4,
        sodium: 0,
        serving_size: '1',
        unitOfMeasure: 'cups'
      },
      {
        name: 'White Rice',
        calories: 68,
        protein: 1.42,
        carbs: 14.84,
        sugar: .03,
        fiber: .2,
        fat: 0.04,
        sodium: 1,
        serving_size: '1/3',
        unitOfMeasure: 'cups'
      },
      {
        name: 'Spaghetti Noodles',
        calories: 200,
        protein: 7,
        carbs: 42,
        sugar: 2,
        fiber: 2,
        fat: 1,
        sodium: 0,
        serving_size: '2',
        unitOfMeasure: 'oz'
      },
      {
        name: 'Classico Pasta Sauce',
        calories: 68,
        protein: 1.42,
        carbs: 14.84,
        sugar: .03,
        fiber: .2,
        fat: 0.04,
        sodium: 1,
        serving_size: '1/3',
        unitOfMeasure: 'cups'
      },
    ]
  }

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <BrowserRouter>
        <QuickPick state={state}/>
      </BrowserRouter>, div)
    ReactDOM.unmountComponentAtNode(div)
  })
  it('renders the UI as exptected', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <QuickPick state={state}/>
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})