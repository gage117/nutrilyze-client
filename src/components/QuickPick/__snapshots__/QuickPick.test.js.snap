// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickPick renders the UI as exptected 1`] = `
<div
  className="quick-pick"
>
  <section
    className="qp__section qp__ingredients"
  >
    <h2
      className="qp__ingredients-header qp__h2"
    >
      Ingredients
    </h2>
    <div
      id="qp__ingredients-container"
    >
      <div
        className="list-div"
      >
        <div
          className="qp__ingredient-bar qp__border-bottom"
          id="Banana"
        >
          <span>
            Banana
          </span>
          <button
            className="Banana qp__add-button"
            id="Banana-button"
            onClick={[Function]}
          >
            Add
          </button>
        </div>
        <div
          className="qp__hidden expand-div qp__border-bottom"
          id="Banana-hidden"
        />
      </div>
      <div
        className="list-div"
      >
        <div
          className="qp__ingredient-bar qp__border-bottom"
          id="Broccoli"
        >
          <span>
            Broccoli
          </span>
          <button
            className="Broccoli qp__add-button"
            id="Broccoli-button"
            onClick={[Function]}
          >
            Add
          </button>
        </div>
        <div
          className="qp__hidden expand-div qp__border-bottom"
          id="Broccoli-hidden"
        />
      </div>
      <div
        className="list-div"
      >
        <div
          className="qp__ingredient-bar qp__border-bottom"
          id="White Rice"
        >
          <span>
            White Rice
          </span>
          <button
            className="White Rice qp__add-button"
            id="White Rice-button"
            onClick={[Function]}
          >
            Add
          </button>
        </div>
        <div
          className="qp__hidden expand-div qp__border-bottom"
          id="White Rice-hidden"
        />
      </div>
      <div
        className="list-div"
      >
        <div
          className="qp__ingredient-bar qp__border-bottom"
          id="Spaghetti Noodles"
        >
          <span>
            Spaghetti Noodles
          </span>
          <button
            className="Spaghetti Noodles qp__add-button"
            id="Spaghetti Noodles-button"
            onClick={[Function]}
          >
            Add
          </button>
        </div>
        <div
          className="qp__hidden expand-div qp__border-bottom"
          id="Spaghetti Noodles-hidden"
        />
      </div>
      <div
        className="list-div"
      >
        <div
          className="qp__ingredient-bar qp__border-bottom"
          id="Classico Pasta Sauce"
        >
          <span>
            Classico Pasta Sauce
          </span>
          <button
            className="Classico Pasta Sauce qp__add-button"
            id="Classico Pasta Sauce-button"
            onClick={[Function]}
          >
            Add
          </button>
        </div>
        <div
          className="qp__hidden expand-div qp__border-bottom"
          id="Classico Pasta Sauce-hidden"
        />
      </div>
    </div>
  </section>
</div>
`;
